editor: Apply Copilot suggestion before fixing indent

mmtftr opened the pull request:
Release Notes:

- Fixed copilot code not getting accepted when tabbed #4549.

mmtftr committed:
editor: apply copilot suggestion before fixing indent #4549

mmtftr commented:
@cla-bot check

as-cii commented:
Thanks for the contribution, @mmtftr! Unfortunately, this breaks the previous behavior that caused Zed to only insert the leading whitespace when hitting `tab` inside of a line's indentation. Can you adjust this PR to fix the issue but also preserve that previous tab behavior?

mmtftr commented:
@as-cii Thanks for the clarification. I don't quite understand the problematic behavior. Is there a test or an issue that I can reference for this behavior? I'd love to write a test to cover this if there isn't any.

mmtftr commented:
So, is the wanted behavior as so:
- If copilot suggestion has lower indent and we have an active suggestion, accept the copilot suggestion even if we're before suggested indent.
- Otherwise, first indent to suggested indent, accept the suggestion on the next tab.

If this is the case, I'll work on implementing this behavior and adding tests for it.

mmtftr committed:
check copilot indent before auto indenting

mmtftr commented:
- Rebased on upstream/main
- Implemented the logic mentioned above `if copilot suggests lower indent than suggested, then avoid auto-indent`
- Wrote tests according to the requirements mentioned above.
  - Also checked locally that the test fails if the editor change is reverted.

mmtftr committed:
simplify condition for copilot precedence over auto indent

mmtftr commented on crates/editor/src/editor.rs:
```diff
@@ -4493,15 +4493,25 @@ impl Editor {
-                {
+                // If the active suggestion implies a lower indent, then don't auto-indent.
+                // Instead, accept the suggestion.
+                && active_suggestion_indent.map_or(true, |ind| {
```
I did the last thing here. I ended up finding another bug with copilot completions but I'll open another issue for that.

mmtftr committed:
remove accidentally committed change

mmtftr commented on crates/copilot_ui/src/copilot_completion_provider.rs:
```diff
@@ -530,6 +530,40 @@ mod tests {
+            &copilot_lsp,
+            vec![copilot::request::Completion {
+                text: "}".into(),
+                range: lsp::Range::new(lsp::Position::new(1, 0), lsp::Position::new(1, 2)),
```
is this range correct?

mmtftr commented on crates/editor/src/editor.rs:
```diff
@@ -4493,15 +4493,31 @@ impl Editor {
+                && cursor.column >= current_indent.len
+                && self.has_active_inline_completion(cx);
+
+            let active_suggestion_indent = if can_accept_suggestion {
```
If this check is not done, we get another issue where
```
fn foo() {
••ˇ• // assume a suggestion exists with proper 4 space indent
}
```
will show a completion but not auto-indent. The problematic behavior is as follows:
1. After first tab
```
fn foo() {
••••ˇ• // editor still shows completion
}
```
2. Second tab will not accept shown suggestion and will instead tab again
```
fn foo() {
••••••••ˇ•
}
```

mmtftr commented on crates/language/src/buffer.rs:
```diff
@@ -3148,7 +3148,7 @@ fn indent_size_for_line(text: &text::BufferSnapshot, row: u32) -> IndentSize {
 }
 
-fn indent_size_for_text(text: impl Iterator<Item = char>) -> IndentSize {
+pub fn indent_size_for_text(text: impl Iterator<Item = char>) -> IndentSize {
```
I exported and used this function directly but I could also add an `impl` function for `Inlay`. Though I don't think it's worth the indirection for a singular use case.

mmtftr commented:
@as-cii could you check this one again?

ConradIrwin commented:
Sorry for the lack of response on this PR.

I am going to close it for now as it conflicts with everything, but I do think this is something we should figure out how to fix. Feel free to re-open and assign me if you're still interested in working on this.
